### PR TITLE
Visualization

### PR DESCRIPTION
**Pull request by jiganerd**
This is a program for visualizing the RoboSail sensor data, based on information sent via the serial port from the Arduino. It runs on Processing, a language/environment made for data visualization, version 1.5.1. 

The main idea is that this could help for debugging and writing code. It could also be useful for demos to advertise the RoboSail program. In the future, it would be neat to be able to visualize the data from the on-the-water RoboSail boats, either in real-time with something like a ZigBee radio, or with data recording/playback.

This has not been tested yet with real RoboSail Arduino output, and in fact there needs to be some minor edits to the current serial output of the Arduino code. (These edits were not made yet in the interest of leaving the Arduino code "untouched" for this pull request.)

visualization

### CODE REVIEW

File: visualization/SensorVisualization/Sailboat.pde [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+    // we also want a magnitude of 50
+    vector.mult(100);
+    
+    windDirection.setVector(vector);
+  }
+  
+  // see notes on sailAngleDeg - here, we calculate the "real" sail angle, 
+  // based on wind direction
+  float getRealSailAngle()
```
Comment (srlm-io) [inline visualization/SensorVisualization/Sailboat.pde:159]: Neat!

File: visualization/SensorVisualization/Sailboat.pde [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    {
+      // not luffing - show white
+      fill(color(240, 240, 240));
+    }
+    pushMatrix();
+    translate(0, 8, 0);
+    
+    // rotate by the sail angle
+    translate(0, 0, 15);
+    rotateY(radians(getRealSailAngle()));
```
Comment (srlm-io) [inline visualization/SensorVisualization/Sailboat.pde]: You have a variable for that above...

File: visualization/SensorVisualization/Sailboat.pde [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+  {
+    float sailHeight = 30;
+    float sailAngleDeg = getRealSailAngle();
+    TruncCone truncCone = new TruncCone();
+
+    pushStyle();
+    if (sailAngleDeg == windAngleDeg)
+    {
+      // the sail is luffing - tint red
+      fill(color(255, 180, 180));
```
Comment (srlm-io) [inline visualization/SensorVisualization/Sailboat.pde:58]: This is a pretty interesting idea. I suppose on the boat side we should never see the sail set to a luff position (at least with the 6 week program as we've taught it), but it's interesting to consider the algorithm for detecting when it's luffing (open loop).

File: visualization/SensorVisualization/SerialParser.pde [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  public float rollDeg;
+  public float pitchDeg;
+  public float yawDeg;
+  public float windAngleDeg;
+  public float rudderAngleDeg;
+  public float sailAngleLimitDeg;
+
+  SerialParser(PApplet parent, String comPort)
+  {
+    sPort = new Serial(parent, comPort, 9600);
```
Comment (srlm-io) [inline visualization/SensorVisualization/SerialParser.pde:29]: Personally I would go with at least 115200. It's reliable and fast(er).

File: visualization/SensorVisualization/SensorVisualization.pde [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+ *
+ *            (overhead view)
+ *                   /                    ^ 90 degrees
+ *            ,,----/-------/             |
+ *      bow  <     +      + |  stern      +----> 0 degrees
+ *            ``------------+             |
+ *                                        v -90 degrees
+ */
+// TODO: better-looking boat
+//       support for Processing 3: textMode(SCREEN), water transparency, axes labelling
```
Comment (srlm-io) [inline visualization/SensorVisualization/SensorVisualization.pde:25]: I noticed that the waves don't accurately reflect the wind state or the boat wake. Pls add. KTHXBYE.

File: visualization/SensorVisualization/RoboSailSerialParser.pde [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  // parse the serial output, in the format of the Arduino program's output, which is
+  // int the format:
+  // 
+  // [joystick x value (0-1023)] [joystick y value (0-1023)] [B1][B2][B3][B4][B5]
+  //
+  // where B1-B1 are 0 or 1 values representing the state of the 5 buttons 
+  // 
+  // this function must be "hooked up" to the global serialEvent() function by the main 
+  // application, e.g. 
+  // void serialEvent(Serial sPort) { serialParser.serialEvent(sPort); }
```
Comment (srlm-io) [inline visualization/SensorVisualization/RoboSailSerialParser.pde:32]: Can you add a bit more on the expected serial format with an example?

As a side note, for systems like these the format is usually just a CSV like `1,34,543.5,3454,34.66,343`, where the full state is output every cycle. It's efficient on the wire and allows the receiver to be stateless (ie, it doesn't have to remember things between sentences, which makes error handling easier).

In other cases it's like NMEA, where you only get partial state for every sentence. UGH.

**Comment by srlm-io** [comment]
This is pretty neat. Overall I think we can go ahead and add this to the repo, and I think it's work playing with more. I'd just like to see a few more documentation notes. I think two things would be helpful here:

1) a README.md with the image, how to run, missing features/important TODOs, etc.
2) An Arduino program that spits out some fake data.

File: visualization/SensorVisualization/RoboSailSerialParser.pde (hunk 6 above)
Comment (jiganerd) [inline visualization/SensorVisualization/RoboSailSerialParser.pde:32]: I was thinking that for simplicity, this program could parse the human-readable serial output from BoatCodeStarter[...].ino - the idea was that the students could use this program to help debug/write code if they want, or they could just look at the text output itself from the Arduino, without having to have some switch in the code to control the type of serial output (or requiring a second serial port). What do you think? (This is also the reason I had the baud rate set to 9600 - it is set that way in BoatCodeStarter[...].ino.) 

There are a few minor edits needed to BoatCodeStarter[...].ino to allow this program to work, though (like for capitalization). Originally, I had made those edits, but backed out of the change as I didn't want to touch the BoatCodeStarter[...].ino without being able to test it. I could, though...

I'm open to ideas...

**Comment by ianfixes** [comment]
I'm preparing to clean up this repo a bit and split it.  This code looks like it should live in a standalone repo under the robosail project group.  To that end, I can set up the README but I will need more information on how to set up my environment to run this program as well as some sample input data (Arduino program not specifically necessary).  

Or am I too late, and is the best thing to just close this issue?

**Comment by jiganerd** [comment]
> I'm preparing to clean up this repo a bit and split it. This code looks like it should live in a standalone repo under the robosail project group. To that end, I can set up the README but I will need more information on how to set up my environment to run this program as well as some sample input data (Arduino program not specifically necessary).
> 
> Or am I too late, and is the best thing to just close this issue?

Hey Ian, cool - for me, it would nice to see this all finalized and an official part of the RoboSail project. I found I had a couple of uncommitted changes on my PC, including a start to a README, that I will push up now. I think I was waiting to have an in-person conversation w/ Cory/srlm-io about the (unfinalized) serial protocol, which never happened.

I am happy to help get this finished and can even come down to the boathouse at some point to try this out w/ the real boats. 

I put some info in the README for now about what's not finished, etc. - I can also supply the Arduino joystick code if that's helpful...

-Brian

File: visualization/SensorVisualization/Sailboat.pde (hunk 2 above)
Comment (jiganerd) [reply visualization/SensorVisualization/Sailboat.pde]: ok, done (3 years later)

**Comment by ianfixes** [comment]
I've created a new repo with the contents of this branch: https://github.com/RoboSail/Visualization

I think future development work (including more visualization projects) can continue in that repository

**Pull request closed without merging**